## Used cars from eBay Kleinanzeigen
Goal:
1. Clean data 
2. Analyze the included used car listings

### 1/9 Import data and take a first look

In [1]:
import pandas as pd
import numpy as np

In [2]:
autos = pd.read_csv('autos.csv', encoding = 'Latin-1')

In [3]:
autos

dateCrawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
5      2016-03-21 13:47:45  Chrysler_Grand_Voyager_2.8_CRD_Aut.Limited_Sto...   
6      2016-03-20 17:55:21  VW_Golf_III_GT_Special_Electronic_Green_Metall...   
7      2016-03-16 18:55:19                               Golf_IV_1.9_TDI_90PS   
8      2016-03-22 16:51:34                                         Seat_Arosa   
9      2016-03-16 13:47:02          Renault_Megane_Scenic_1.6e_RT_Klimaanlage   
10     2016-03-15 01:41:36                       VW_Golf_Tuning_in_siber/grau   
11     2016-03-16 18:45:34                         Mercedes_A140_Motorschaden   
12     2016-03-31 19:48:22  Smart_smart_fortwo_coupe_softouch_pure_MHD_Pan...   
13     2016-03-23 10:48:32                                 Audi_A3_1.6_tuning   
14     2016-03-23 11:50:46  Renault_Clio_3__Dynamique_1.2__16_V;_viele_Ver...   
15     2016-04-01 12:06:20        Corvette_C3_Coupe_T_Top_Crossfire_Injection   
16     2016-03-16 14:59:02                                Opel_Vectra_B_Kombi   
17     2016-03-29 11:46:22                          Volkswagen_Scirocco_2_G60   
18     2016-03-26 19:57:44               Verkaufen_mein_bmw_e36_320_i_touring   
19     2016-03-17 13:36:21        mazda_tribute_2.0_mit_gas_und_tuev_neu_2018   
20     2016-03-05 19:57:31  Audi_A4_Avant_1.9_TDI_*6_Gang*AHK*Klimatronik*...   
21     2016-03-06 19:07:10                      Porsche_911_Carrera_4S_Cabrio   
22     2016-03-28 20:50:54                               MINI_Cooper_S_Cabrio   
23     2016-03-10 19:55:34   Peugeot_Boxer_2_2_HDi_120_Ps_9_Sitzer_inkl_Klima   
24     2016-04-03 11:57:02                         BMW_535i_xDrive_Sport_Aut.   
25     2016-03-21 21:56:18  Ford_escort_kombi_an_bastler_mit_ghia_ausstattung   
26     2016-04-03 22:46:28                                Volkswagen_Polo_Fox   
27     2016-03-27 18:45:01             Hat_einer_Ahnung_mit_Ford_Galaxy_HILFE   
28     2016-03-19 21:56:19                                      MINI_Cooper_D   
29     2016-04-02 12:45:44  Mercedes_Benz_E_320_T_CDI_Avantgarde_DPF7_Sitz...   
...                    ...                                                ...   
49970  2016-03-21 22:47:37  c4_Grand_Picasso_mit_Automatik_Leder_Navi_Temp...   
49971  2016-03-29 14:54:12                                         W.Lupo_1.0   
49972  2016-03-26 22:25:23          Mercedes_Benz_Vito_115_CDI_Extralang_Aut.   
49973  2016-03-27 05:32:39                   Mercedes_Benz_SLK_200_Kompressor   
49974  2016-03-20 10:52:31  Golf_1_Cabrio_Tuev_Neu_viele_Extras_alles_eing...   
49975  2016-03-27 20:51:39          Honda_Jazz_1.3_DSi_i_VTEC_IMA_CVT_Comfort   
49976  2016-03-19 18:56:05  Audi_80_Avant_2.6_E__Vollausstattung!!_Einziga...   
49977  2016-03-31 18:37:18                        Mercedes_Benz_C200_Cdi_W203   
49978  2016-04-04 10:37:14                        Mercedes_Benz_E_200_Classic   
49979  2016-03-20 18:38:40                      Volkswagen_Polo_1.6_TDI_Style   
49980  2016-03-12 10:55:54                            Ford_Escort_Turnier_16V   
49981  2016-03-15 09:38:21             Opel_Astra_Kombi_mit_Anhaengerkupplung   
49982  2016-03-29 18:51:08            Skoda_Fabia_4_Tuerer_Bj:2004__85.000Tkm   
49983  2016-03-06 12:43:04                                      Ford_focus_99   
49984  2016-03-31 22:48:48  Student_sucht_ein__Anfaengerauto___ab_2000_BJ_...   
49985  2016-04-02 16:38:23                          Verkaufe_meinen_vw_vento!   
49986  2016-04-04 20:46:02  Chrysler_300C_3.0_CRD_DPF_Automatik_Voll_Ausst...   
49987  2016-03-22 20:47:27  Audi_A3_Limousine_

In [4]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
dateCrawled            50000 non-null object
name                   50000 non-null object
seller                 50000 non-null object
offerType              50000 non-null object
price                  50000 non-null object
abtest                 50000 non-null object
vehicleType            44905 non-null object
yearOfRegistration     50000 non-null int64
gearbox                47320 non-null object
powerPS                50000 non-null int64
model                  47242 non-null object
odometer               50000 non-null object
monthOfRegistration    50000 non-null int64
fuelType               45518 non-null object
brand                  50000 non-null object
notRepairedDamage      40171 non-null object
dateCreated            50000 non-null object
nrOfPictures           50000 non-null int64
postalCode             50000 non-null int64
lastSeen               50000 non-null obj

In [5]:
autos.head(10)

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
5  2016-03-21 13:47:45  Chrysler_Grand_Voyager_2.8_CRD_Aut.Limited_Sto...   
6  2016-03-20 17:55:21  VW_Golf_III_GT_Special_Electronic_Green_Metall...   
7  2016-03-16 18:55:19                               Golf_IV_1.9_TDI_90PS   
8  2016-03-22 16:51:34                                         Seat_Arosa   
9  2016-03-16 13:47:02          Renault_Megane_Scenic_1.6e_RT_Klimaanlage   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   
5  privat   Angebot  $7,900     test         bus                2006   
6  privat   Angebot    $300     test   limousine                1995   
7  privat   Angebot  $1,990  control   limousine                1998   
8  privat   Angebot    $250     test         NaN                2000   
9  privat   Angebot    $590  control         bus                1997   

     gearbox  powerPS    model   odometer  monthOfRegistration fuelType  \
0    manuell      158   andere  150,000km                    3      lpg   
1  automatik      286      7er  150,000km                    6   benzin   
2    manuell      102     golf   70,000km                    7   benzin   
3  automatik       71   fortwo   70,000km                    6   benzin   
4    manuell        0    focus  150,000km                    7   benzin   
5  automatik      150  voyager  150,000km                    4   diesel   
6    manuell       90     golf  150,000km                    8   benzin   
7    manuell       90     golf  150,000km                   12   diesel   
8    manuell        0    arosa  150,000km                   10      NaN   
9    manuell       90   megane  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   
5    chrysler               NaN  2016-03-21 00:00:00             0   
6  volkswagen               NaN  2016-03-20 00:00:00             0   
7  volkswagen              nein  2016-03-16 00:00:00             0   
8        seat              nein  2016-03-22 00:00:00             0   
9     renault              nein  2016-03-16 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50  
5       22962  2016-04-06 09:45:21  
6       31535  2016-03-23 02:48:59  
7       53474  2016-04-07 03:17:32  
8        7426  2016-03-26 18:18:10  
9       15749  2016-04-06 10:46:35

My observation:

1. Sample size 50000*20
2. There are 5 features that contain null values: vehicleType, gearbox, model, fuelType, notRepairedDamage
3. Some values are in German 

DQ observation: 

1. The dataset contains 20 columns, most of which are strings.
2. Some columns have null values, but none have more than ~20% null values.
3. The column names use camelcase instead of Python's preferred snakecase, which means we can't just replace spaces with underscores.

### 2/9 Change column names

In [6]:
# Change names to a few columns:
autos.rename({'yearOfRegistration':'registration_year', 'monthOfRegistration':'registration_month', 'notRepairedDamage':'unrepaired_damage', 'dateCreated':'ad_created'}, axis = 1, inplace = True)

In [7]:
# Convert camelCasing to snake_casing in the rest of the columns 
columns = autos.columns 
new_columns = columns.str.replace('(?<!^)(?=[A-Z])', '_').str.lower()
autos.columns = new_columns

In [8]:
print(autos.columns)

Index(['date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest',
       'vehicle_type', 'registration_year', 'gearbox', 'power_p_s', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'nr_of_pictures', 'postal_code',
       'last_seen'],
      dtype='object')


In [9]:
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price   abtest vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

     gearbox  power_p_s   model   odometer  registration_month fuel_type  \
0    manuell        158  andere  150,000km                   3       lpg   
1  automatik        286     7er  150,000km                   6    benzin   
2    manuell        102    golf   70,000km                   7    benzin   
3  automatik         71  fortwo   70,000km                   6    benzin   
4    manuell          0   focus  150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  nr_of_pictures  \
0     peugeot              nein  2016-03-26 00:00:00               0   
1         bmw              nein  2016-04-04 00:00:00               0   
2  volkswagen              nein  2016-03-26 00:00:00               0   
3       smart              nein  2016-03-12 00:00:00               0   
4        ford              nein  2016-04-01 00:00:00               0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

Changes made so far: 

1. Changes the names of a few columns to be more concise
2. Convert rest of the column names from camelCasing to snake_casing

### 3/9 Explore all columns with describe method

In [10]:
# Take a look at the actual data
autos.describe(include = 'all') # Includes both numerical and non-numerical data

date_crawled         name  seller offer_type  price abtest  \
count                 50000        50000   50000      50000  50000  50000   
unique                48213        38754       2          2   2357      2   
top     2016-03-22 09:51:06  Ford_Fiesta  privat    Angebot     $0   test   
freq                      3           78   49999      49999   1421  25756   
mean                    NaN          NaN     NaN        NaN    NaN    NaN   
std                     NaN          NaN     NaN        NaN    NaN    NaN   
min                     NaN          NaN     NaN        NaN    NaN    NaN   
25%                     NaN          NaN     NaN        NaN    NaN    NaN   
50%                     NaN          NaN     NaN        NaN    NaN    NaN   
75%                     NaN          NaN     NaN        NaN    NaN    NaN   
max                     NaN          NaN     NaN        NaN    NaN    NaN   

       vehicle_type  registration_year  gearbox     power_p_s  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  nr_of_pictures   postal_code            last_seen  
count                 50000         50000.0  50000.000000                50000  
unique                   76             NaN           NaN                39481  
top     2016-04-03 00:00:00             NaN           NaN  2016-04-07 06:17:27  
freq                   1946             NaN           NaN                    8  
mean                    NaN             0.0  50813.627300                  NaN  
std                     NaN             0.0  25779.747957                  NaN  
min                     NaN             0.0   1067.000000                  NaN  
25%                     NaN             0.0  30451.000000                  NaN  
50%                     NaN             0.0  49577.000000                  NaN  
75%                     NaN             0.0  71540.000000                  NaN  
max                     NaN             0.0  99998.000000                  NaN

1. Columns 'seller' and 'offer_type' have mostly same values, which means they can be dropped. 
2. Investigation required abnormalities:
   1. Column 'price' has more than a thousand zeros
   2. 'registration_year' has a max of 9999
   3. 'registration_month' month has a min of 0
   4. 'nr_of_pictures' has all zeros 
   5. 'postal_code' max and min digits are inconsistant
3. Non-numerical columns 'price', 'odometer' need to be converted to numerical columns 
   

In [11]:
autos[autos['price'] == '$0' ]

date_crawled                                               name  \
27     2016-03-27 18:45:01             Hat_einer_Ahnung_mit_Ford_Galaxy_HILFE   
71     2016-03-28 19:39:35  Suche_Opel_Astra_F__Corsa_oder_Kadett_E_mit_Re...   
80     2016-03-09 15:57:57  Nissan_Primera_Hatchback_1_6_16v_73_Kw___99Ps_...   
87     2016-03-29 23:37:22                      Bmw_520_e39_zum_ausschlachten   
99     2016-04-05 09:48:54                    Peugeot_207_CC___Cabrio_Bj_2011   
118    2016-03-12 05:03:00  VW_Sharan_V6_204_PS_Karosse_Rohkarosse_mit_Pap...   
146    2016-03-22 23:59:28                                    Ford_Fiesta_rot   
167    2016-04-02 19:43:45  Suche_VW_Multivan_Innenausstattung_Set_oder_TE...   
180    2016-03-19 10:50:25                                       Zu_verkaufen   
226    2016-03-25 23:52:12                      Porsche_911_S_Targa__67er_SWB   
234    2016-03-05 16:56:22                                   Fiat_Punto_Sport   
248    2016-03-08 22:57:53                             VW_Passat_zu_verkaufen   
259    2016-04-03 23:49:58                      guenstiges_Auto_/_auch_defekt   
301    2016-03-08 20:37:59                         Kaufe_alle_Autos_bietet_an   
323    2016-03-15 16:54:59                                W220_Unfallfahrzeug   
327    2016-04-03 13:49:21              BMW_Cabrio_318i_/_BMW_318_tds_Touring   
347    2016-04-02 11:48:02                                  Volvo_TP21__Sugga   
366    2016-03-21 20:36:25  Ford_Escort_Express_75__LKW_Zulussung___Diesel...   
389    2016-03-08 11:38:52                                    Vw_polo_classic   
418    2016-03-29 14:43:24         Fiat_SCUDO_8_Sitzer_Bus__Diesel_JTD__80_KW   
427    2016-03-12 11:44:47                 TOP_ZUSTAND_GrunePlakette_Euro4!!!   
430    2016-03-18 23:52:40                                  Winterraeder_FORD   
449    2016-04-03 20:59:59                           Tausche_gegen_ein_smart.   
487    2016-03-14 21:41:36  MAG_MICH_KEINER_HABEN_Opel_Omega_B_Caravan_2.0...   
502    2016-03-25 23:55:24                               C32_AMG_C55_C43_w203   
519    2016-03-28 08:53:58  Verkaufe/Tausche_Seat_Inca__VW_Caddy__aus_2002...   
565    2016-03-20 20:49:36     Tausche_Audi_S3_8p_gegen_Golf_VI_Gti_mit_xenon   
593    2016-04-01 21:44:15                              CLK_200__BRABUS_Optik   
660    2016-03-29 21:44:40                                       Opel_Corsa_B   
661    2016-03-18 22:49:48               Renault_Megane_Scenic_1.6_Tuev_12/17   
...                    ...                                                ...   
48882  2016-03-14 23:58:47                                         SUCHE_AUTO   
48889  2016-04-01 10:50:52                           Audi_A4_____TueV_09.2016   
48891  2016-03-20 22:50:14        MB_E_220_CDI__TAUSCHEN_nur_SUV/GELÄNDEWAGEN   
48982  2016-03-08 12:48:33                           Bmw_525D_Sport_Pro..Navi   
48985  2016-03-27 08:55:25                             Stockcar_Autos_Peugeot   
49043  2016-03-07 11:48:55                              Opel_Vectra_B_Caravan   
49082  2016-03-13 17:06:58         Passat_35i_b4_in_candy_weiss_original_lack   
49254  2016-03-21 10:49:57                           Corrado_G60_mit_2e_Motor   
49301  2016-03-20 20:52:40                           Golf_3_Variant_Kombi_1.8   
49342  2016-03-09 20:47:01  Audi_80_b4_S_Color_und_BMW_318ti_compact_Tausc...   
49350  2016-03-19 23:57:15                                     Golf_II_Diesel   
49404  2016-04-04 03:03:16                                Golf_2_BBS_LEDER_KW   
49408  2016-03-30 20:57:06                                         OPEL_ASTRA   
49479  2016-04-01 23:55:01                            Tausch_mit_frischen_TÜV   
49496  2016-03-26 13:55:28                                        Bmw_e39_520   
49504  2016-03-13 19:42:57                                 Meriva_Lpg_1_6_16v   
49507  2016-03-16 17:38:32                 BMW_528i_E39_Schaltgetriebe_Unfall   
49525  2016-03-09 17:59:02              Bmw_e

In [12]:
# Convert date_crawled column to datetime
autos['date_crawled'] = pd.to_datetime(autos['date_crawled'] ) 
most_recent = autos['date_crawled'].max()
least_recent = autos['date_crawled'].min()
print(most_recent, least_recent)

2016-04-07 14:36:56 2016-03-05 14:06:30


In [13]:
# Realistically registration year shouldn't be bigger than crawled date
autos[autos['registration_year']>2016].info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1966 entries, 10 to 49935
Data columns (total 20 columns):
date_crawled          1966 non-null datetime64[ns]
name                  1966 non-null object
seller                1966 non-null object
offer_type            1966 non-null object
price                 1966 non-null object
abtest                1966 non-null object
vehicle_type          2 non-null object
registration_year     1966 non-null int64
gearbox               1715 non-null object
power_p_s             1966 non-null int64
model                 1680 non-null object
odometer              1966 non-null object
registration_month    1966 non-null int64
fuel_type             1215 non-null object
brand                 1966 non-null object
unrepaired_damage     1129 non-null object
ad_created            1966 non-null object
nr_of_pictures        1966 non-null int64
postal_code           1966 non-null int64
last_seen             1966 non-null object
dtypes: datetime64[ns](1), int6

In [14]:
autos['registration_month'].value_counts() # 5075 entires of unusual value month-0

0     5075
3     5071
6     4368
5     4107
4     4102
7     3949
10    3651
12    3447
9     3389
11    3360
1     3282
8     3191
2     3008
Name: registration_month, dtype: int64

In [15]:
autos['nr_of_pictures'].value_counts() # This columns can be dropped with all zeros 

0    50000
Name: nr_of_pictures, dtype: int64

In [16]:
autos['postal_code']

0        79588
1        71034
2        35394
3        33729
4        39218
5        22962
6        31535
7        53474
8         7426
9        15749
10       86157
11       17498
12       34590
13       12043
14       81737
15       61276
16       57299
17       74821
18       54329
19       40878
20       53913
21       65428
22       44789
23       30900
24       22547
25       27574
26       38110
27       66701
28       15745
29       47638
         ...  
49970    14947
49971    65197
49972    65326
49973    53567
49974     8209
49975    84385
49976    36100
49977    33739
49978    24405
49979    48455
49980    56218
49981    86859
49982    45884
49983    52477
49984    12103
49985    30900
49986    73527
49987    94362
49988    83646
49989    21244
49990    58239
49991    61350
49992    68642
49993    65203
49994    46537
49995    82131
49996    44807
49997    73430
49998    35683
49999    45897
Name: postal_code, Length: 50000, dtype: int64

In [17]:
autos.loc[autos['postal_code'] < 1e4, 'postal_code']

8        7426
40       6780
54       6886
60       1665
71       4552
79       4157
150      1259
215      6628
313      6571
361      6628
365      1259
366      3130
375      4610
393      9111
395      6886
396      2828
403      7318
415      6114
428      6258
446      1594
453      7545
464      6842
484      4316
524      1920
533      3229
539      6712
590      4317
612      6679
617      6712
648      2994
         ... 
49171    9456
49173    1968
49213    9669
49228    1445
49235    9322
49251    1159
49278    4758
49301    4924
49304    2906
49306    4626
49365    7819
49393    7548
49410    6120
49414    6130
49465    7330
49473    3042
49480    1157
49551    4275
49581    1594
49609    1768
49632    6844
49698    9366
49743    8141
49748    8066
49789    4639
49822    2692
49906    7743
49925    4316
49944    4105
49974    8209
Name: postal_code, Length: 2428, dtype: int64

In [18]:
autos.loc[autos['postal_code'] < 1e4, 'postal_code'].count() 
# 2428 entries with 4 digits postal code

2428

In [19]:
# Convert non-numerical columns 'price' into numerical columns
autos['price']

0         $5,000
1         $8,500
2         $8,990
3         $4,350
4         $1,350
5         $7,900
6           $300
7         $1,990
8           $250
9           $590
10          $999
11          $350
12        $5,299
13        $1,350
14        $3,999
15       $18,900
16          $350
17        $5,500
18          $300
19        $4,150
20        $3,500
21       $41,500
22       $25,450
23        $7,999
24       $48,500
25           $90
26          $777
27            $0
28        $5,250
29        $4,999
          ...   
49970    $15,800
49971       $950
49972     $3,300
49973     $6,000
49974         $0
49975     $9,700
49976     $5,900
49977     $5,500
49978       $900
49979    $11,000
49980       $400
49981     $2,000
49982     $1,950
49983       $600
49984         $0
49985     $1,000
49986    $15,900
49987    $21,990
49988     $9,550
49989       $150
49990    $17,500
49991       $500
49992     $4,800
49993     $1,650
49994     $5,000
49995    $24,900
49996     $1,980
49997    $13,2

In [20]:
autos['price'] = autos['price'].str.replace('\D', '') # Remove all non-digits(/D)

In [21]:
autos['price'] = autos['price'].astype(int) # Digitize
autos['price'].dtypes

dtype('int64')

In [22]:
# Convert non-numerical columns 'odometer' into numerical columns
autos['odometer']

0        150,000km
1        150,000km
2         70,000km
3         70,000km
4        150,000km
5        150,000km
6        150,000km
7        150,000km
8        150,000km
9        150,000km
10       150,000km
11       150,000km
12        50,000km
13       150,000km
14       150,000km
15        80,000km
16       150,000km
17       150,000km
18       150,000km
19       150,000km
20       150,000km
21       150,000km
22        10,000km
23       150,000km
24        30,000km
25       150,000km
26       125,000km
27       150,000km
28       150,000km
29       150,000km
           ...    
49970     60,000km
49971    150,000km
49972    150,000km
49973    150,000km
49974    150,000km
49975    100,000km
49976    150,000km
49977    150,000km
49978    150,000km
49979     70,000km
49980    125,000km
49981    150,000km
49982     90,000km
49983    150,000km
49984    150,000km
49985    150,000km
49986    125,000km
49987     50,000km
49988    150,000km
49989    150,000km
49990     30,000km
49991    150

In [23]:
autos['odometer'] = autos['odometer'].str.replace('\D', '')
autos['odometer'].astype(int)

0        150000
1        150000
2         70000
3         70000
4        150000
5        150000
6        150000
7        150000
8        150000
9        150000
10       150000
11       150000
12        50000
13       150000
14       150000
15        80000
16       150000
17       150000
18       150000
19       150000
20       150000
21       150000
22        10000
23       150000
24        30000
25       150000
26       125000
27       150000
28       150000
29       150000
          ...  
49970     60000
49971    150000
49972    150000
49973    150000
49974    150000
49975    100000
49976    150000
49977    150000
49978    150000
49979     70000
49980    125000
49981    150000
49982     90000
49983    150000
49984    150000
49985    150000
49986    125000
49987     50000
49988    150000
49989    150000
49990     30000
49991    150000
49992    125000
49993    150000
49994    150000
49995    100000
49996    150000
49997      5000
49998     40000
49999    150000
Name: odometer, Length: 

In [24]:
autos.rename({'odometer':'odometer_km'}, inplace = True, axis = 1) # Rename 'odometer' column
autos.head()

date_crawled                                               name  \
0 2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1 2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2 2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3 2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4 2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type  price   abtest vehicle_type  registration_year  \
0  privat    Angebot   5000  control          bus               2004   
1  privat    Angebot   8500  control    limousine               1997   
2  privat    Angebot   8990     test    limousine               2009   
3  privat    Angebot   4350  control   kleinwagen               2007   
4  privat    Angebot   1350     test        kombi               2003   

     gearbox  power_p_s   model odometer_km  registration_month fuel_type  \
0    manuell        158  andere      150000                   3       lpg   
1  automatik        286     7er      150000                   6    benzin   
2    manuell        102    golf       70000                   7    benzin   
3  automatik         71  fortwo       70000                   6    benzin   
4    manuell          0   focus      150000                   7    benzin   

        brand unrepaired_damage           ad_created  nr_of_pictures  \
0     peugeot              nein  2016-03-26 00:00:00               0   
1         bmw              nein  2016-04-04 00:00:00               0   
2  volkswagen              nein  2016-03-26 00:00:00               0   
3       smart              nein  2016-03-12 00:00:00               0   
4        ford              nein  2016-04-01 00:00:00               0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

### 4/9 Explore 'odometer_km' and 'price' columns 

In [25]:
# Check out 'price' values, find&drop outliers
autos['price'].unique().shape , autos['odometer_km'].unique().shape

((2357,), (13,))

In [26]:
autos['price'].describe()

count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64

In [27]:
autos['price'].value_counts().sort_index(ascending = True).head(20)

0     1421
1      156
2        3
3        1
5        2
8        1
9        1
10       7
11       2
12       3
13       2
14       1
15       2
17       3
18       1
20       4
25       5
29       1
30       7
35       1
Name: price, dtype: int64

In [28]:
autos['price'].value_counts().sort_index(ascending = False).head(20)

99999999    1
27322222    1
12345678    3
11111111    2
10000000    1
3890000     1
1300000     1
1234566     1
999999      2
999990      1
350000      1
345000      1
299000      1
295000      1
265000      1
259000      1
250000      1
220000      1
198000      1
197000      1
Name: price, dtype: int64

In [29]:
autos.loc[(autos['price'] > 1000) & (autos['price'] < 60000), 'price'].count()

37862

1.  'price' has a std of 4.811044e+05 (with biased values)
2. After trying out with different conditions on how many items there are between prices,also based on the reality of price of used cars, dropping rows with prices outside of range 1000-60000

In [30]:
# Drop all rows with a price outside of range 1000-600000
autos.drop(((autos[autos['price']<1000])|(autos[autos['price']>60000])).index, inplace = True)
autos.describe() # Showing only numerical columns without include = 'all' param

price  registration_year     power_p_s  registration_month  \
count  38504.000000       38504.000000  38504.000000        38504.000000   
mean    6944.071395        2005.680033    128.222626            5.998208   
std     7240.796479          86.818848    215.733910            3.592504   
min     1000.000000        1000.000000      0.000000            0.000000   
25%     2200.000000        2001.000000     80.000000            3.000000   
50%     4300.000000        2005.000000    116.000000            6.000000   
75%     8900.000000        2009.000000    160.000000            9.000000   
max    60000.000000        9999.000000  17700.000000           12.000000   

       nr_of_pictures   postal_code  
count         38504.0  38504.000000  
mean              0.0  51673.562565  
std               0.0  25686.923356  
min               0.0   1067.000000  
25%               0.0  31198.000000  
50%               0.0  50827.000000  
75%               0.0  72517.000000  
max               0.0  99998.000000

In [31]:
# Check out 'odometer_km' values, find&drop outliers
autos['odometer_km'].describe()

count      38504
unique        13
top       150000
freq       23306
Name: odometer_km, dtype: object

In [32]:
autos['odometer_km'].value_counts()

150000    23306
125000     4332
100000     1856
90000      1566
80000      1332
70000      1152
60000      1087
50000       980
40000       783
30000       728
20000       672
5000        492
10000       218
Name: odometer_km, dtype: int64

No outliers found for 'odometer_km' column

### 5/9 Explore date related columns 

In [33]:
# 'date_crawled' was converted to type Datetime before
autos['date_crawled'].value_counts()[:10]

2016-04-04 16:40:33    3
2016-04-02 11:37:04    3
2016-03-23 19:38:20    3
2016-03-19 17:36:18    3
2016-03-09 11:54:38    3
2016-03-08 10:40:35    3
2016-03-29 23:42:13    3
2016-03-12 14:42:09    2
2016-04-01 14:49:35    2
2016-03-26 10:37:56    2
Name: date_crawled, dtype: int64

In [34]:
autos['date_crawled'].dt.date.value_counts(dropna = False, normalize = True).sort_index() 
# Above returns the counts of date value counts in 'date_crawled' columns in percentage with ascending dates
# Series.dt.date 
#Returns numpy array of python datetime.date objects (namely, the date part of Timestamps without timezone information).

2016-03-05    0.025530
2016-03-06    0.013895
2016-03-07    0.035113
2016-03-08    0.032620
2016-03-09    0.032516
2016-03-10    0.033399
2016-03-11    0.032776
2016-03-12    0.037425
2016-03-13    0.016050
2016-03-14    0.036646
2016-03-15    0.033607
2016-03-16    0.029062
2016-03-17    0.030464
2016-03-18    0.012804
2016-03-19    0.035191
2016-03-20    0.038178
2016-03-21    0.037269
2016-03-22    0.032490
2016-03-23    0.032178
2016-03-24    0.029036
2016-03-25    0.030542
2016-03-26    0.033113
2016-03-27    0.031347
2016-03-28    0.035373
2016-03-29    0.034022
2016-03-30    0.033062
2016-03-31    0.031399
2016-04-01    0.034568
2016-04-02    0.036282
2016-04-03    0.039113
2016-04-04    0.036827
2016-04-05    0.013323
2016-04-06    0.003272
2016-04-07    0.001506
Name: date_crawled, dtype: float64

This data is crawled between 2016/03/05 and 2016/04/07

*Note that this is not done as required because the 'date_crawled' column was converted from string to datetime in ealier data exploration. But the other two date related columns will be explored with mathods for string columns below.

In [35]:
autos['ad_created'].str[:10].value_counts(normalize = True, dropna = False).sort_index()
# Sorts the first 10 digits(date only) value counts in 'ad_created' column in ascending order with percentage

2015-06-11    0.000026
2015-08-10    0.000026
2015-09-09    0.000026
2015-11-10    0.000026
2015-12-30    0.000026
2016-01-03    0.000026
2016-01-07    0.000026
2016-01-10    0.000052
2016-01-13    0.000026
2016-01-14    0.000026
2016-01-16    0.000026
2016-01-22    0.000026
2016-01-27    0.000078
2016-01-29    0.000026
2016-02-01    0.000026
2016-02-02    0.000052
2016-02-05    0.000052
2016-02-07    0.000026
2016-02-09    0.000026
2016-02-11    0.000026
2016-02-12    0.000052
2016-02-14    0.000052
2016-02-16    0.000026
2016-02-17    0.000026
2016-02-18    0.000052
2016-02-19    0.000078
2016-02-20    0.000026
2016-02-21    0.000052
2016-02-22    0.000026
2016-02-23    0.000104
                ...   
2016-03-09    0.032724
2016-03-10    0.033062
2016-03-11    0.033036
2016-03-12    0.037165
2016-03-13    0.017712
2016-03-14    0.034957
2016-03-15    0.033451
2016-03-16    0.029633
2016-03-17    0.030127
2016-03-18    0.013271
2016-03-19    0.034126
2016-03-20    0.038282
2016-03-21 

In [36]:
autos['ad_created'].str[:10].value_counts(normalize = True, dropna = False).max()

0.03939850405152712

1. Ads are created between 2015/06/11 and 2016/04/07
2. Most ads are created in 2016

In [37]:
autos['last_seen'].str[:10].value_counts(normalize = True, dropna = False).sort_index()
# Sorts the first 10 digits(date only) value counts in 'last_seen' column in ascending order with percentage

2016-03-05    0.001091
2016-03-06    0.003584
2016-03-07    0.004519
2016-03-08    0.006259
2016-03-09    0.008882
2016-03-10    0.009843
2016-03-11    0.011765
2016-03-12    0.022205
2016-03-13    0.008415
2016-03-14    0.012025
2016-03-15    0.015011
2016-03-16    0.015453
2016-03-17    0.026465
2016-03-18    0.007350
2016-03-19    0.014622
2016-03-20    0.019790
2016-03-21    0.019712
2016-03-22    0.020829
2016-03-23    0.017920
2016-03-24    0.018518
2016-03-25    0.017816
2016-03-26    0.016050
2016-03-27    0.014050
2016-03-28    0.019427
2016-03-29    0.020829
2016-03-30    0.023478
2016-03-31    0.022751
2016-04-01    0.023244
2016-04-02    0.024932
2016-04-03    0.024413
2016-04-04    0.023452
2016-04-05    0.130844
2016-04-06    0.234625
2016-04-07    0.139830
Name: last_seen, dtype: float64

1. Ads are last seen between 2016/03/05 and 2016/04/07
2. More than half are last seen in April 2016

### 6/9 Explore 'registration_year' column

In [38]:
autos['registration_year'].describe()

count    38504.000000
mean      2005.680033
std         86.818848
min       1000.000000
25%       2001.000000
50%       2005.000000
75%       2009.000000
max       9999.000000
Name: registration_year, dtype: float64

The 'registration_year' columns has been explored a little earlier. Realistically registration year shouldn't be bigger than crawled date or smaller than car invention.

In [39]:
autos['registration_year'].value_counts().sort_index()

1000       1
1001       1
1927       1
1929       1
1931       1
1934       2
1937       4
1938       1
1939       1
1941       2
1943       1
1948       1
1950       1
1951       1
1952       1
1953       1
1954       2
1955       1
1956       4
1957       2
1958       3
1959       6
1960      17
1961       6
1962       4
1963       7
1964       8
1965      17
1966      21
1967      25
        ... 
1999    1724
2000    2000
2001    2064
2002    2135
2003    2477
2004    2606
2005    2785
2006    2650
2007    2254
2008    2199
2009    2073
2010    1577
2011    1610
2012    1298
2013     780
2014     638
2015     343
2016     644
2017    1008
2018     396
2019       1
2800       1
4100       1
4500       1
5000       2
5911       1
6200       1
8888       1
9000       1
9999       2
Name: registration_year, Length: 91, dtype: int64

In [40]:
autos['registration_year'].describe()

count    38504.000000
mean      2005.680033
std         86.818848
min       1000.000000
25%       2001.000000
50%       2005.000000
75%       2009.000000
max       9999.000000
Name: registration_year, dtype: float64

In [41]:
autos.loc[autos['registration_year'] < 2017, 'registration_year'].describe()
# We are sure that any data out side of these 37088 entries are inaccurate because 
# registration year can't be bigger than crawled date.

count    37088.000000
mean      2003.910267
std         10.153671
min       1000.000000
25%       2001.000000
50%       2005.000000
75%       2008.000000
max       2016.000000
Name: registration_year, dtype: float64

In [42]:
autos.loc[autos['registration_year'] < 1970, 'registration_year'].describe()

count     186.000000
mean     1951.693548
std        99.809915
min      1000.000000
25%      1960.000000
50%      1965.000000
75%      1967.000000
max      1969.000000
Name: registration_year, dtype: float64

After looking at amount of listings in the 1990s, year 1970 seems like a good begining year.

In [43]:
autos.loc[(autos['registration_year'] < 2017) & (autos['registration_year'] > 1970), 'registration_year'].describe()

count    36870.000000
mean      2004.203119
std          6.235416
min       1971.000000
25%       2001.000000
50%       2005.000000
75%       2008.000000
max       2016.000000
Name: registration_year, dtype: float64

Between year 1970-2016 inclusive, the remaining 36870 entries looks realistic.

In [44]:
autos.drop(autos.loc[(autos['registration_year'] > 2016) | (autos['registration_year'] < 1970)].index, inplace = True)
autos['registration_year'].value_counts(normalize = True)

2005    0.075470
2006    0.071812
2004    0.070619
2003    0.067124
2007    0.061081
2008    0.059590
2002    0.057856
2009    0.056176
2001    0.055932
2000    0.054198
1999    0.046718
2011    0.043629
2010    0.042735
2012    0.035174
1998    0.034659
1997    0.021245
2013    0.021137
2016    0.017452
2014    0.017289
1996    0.014633
1995    0.011951
2015    0.009295
1994    0.007344
1993    0.005880
1992    0.005853
1991    0.005582
1990    0.005067
1989    0.003333
1988    0.002845
1985    0.002141
1980    0.001897
1986    0.001518
1983    0.001355
1987    0.001355
1984    0.001274
1978    0.001111
1982    0.001030
1970    0.000867
1979    0.000840
1972    0.000813
1981    0.000705
1973    0.000623
1974    0.000623
1971    0.000596
1976    0.000569
1977    0.000515
1975    0.000488
Name: registration_year, dtype: float64

1. More than 90% of the listings are between year 1990-2016
2. More than 60% of the listings are between 2000-2009
3. Less than 1% of the listings are from the 1970s so probably should drop the 1970s data

In [45]:
autos.drop(autos.loc[autos['registration_year'] < 1980].index, inplace = True)
autos['registration_year'].value_counts(normalize = True)

2005    0.076006
2006    0.072321
2004    0.071121
2003    0.067600
2007    0.061514
2008    0.060013
2002    0.058266
2009    0.056574
2001    0.056329
2000    0.054582
1999    0.047050
2011    0.043939
2010    0.043038
2012    0.035424
1998    0.034905
1997    0.021396
2013    0.021287
2016    0.017575
2014    0.017412
1996    0.014737
1995    0.012035
2015    0.009361
1994    0.007396
1993    0.005922
1992    0.005895
1991    0.005622
1990    0.005103
1989    0.003357
1988    0.002866
1985    0.002156
1980    0.001910
1986    0.001528
1983    0.001365
1987    0.001365
1984    0.001283
1982    0.001037
1981    0.000710
Name: registration_year, dtype: float64

### 7/9 Explore 'brand' column

In [46]:
autos['brand'].unique()

array(['peugeot', 'bmw', 'volkswagen', 'smart', 'ford', 'chrysler',
       'audi', 'renault', 'sonstige_autos', 'mazda', 'porsche', 'mini',
       'mercedes_benz', 'seat', 'toyota', 'opel', 'dacia', 'jeep', 'saab',
       'volvo', 'nissan', 'jaguar', 'skoda', 'subaru', 'fiat',
       'mitsubishi', 'chevrolet', 'hyundai', 'honda', 'kia', 'citroen',
       'suzuki', 'land_rover', 'alfa_romeo', 'rover', 'trabant',
       'daihatsu', 'daewoo', 'lancia', 'lada'], dtype=object)

In [47]:
autos['brand'].value_counts(normalize = True)

volkswagen        0.212325
bmw               0.126521
mercedes_benz     0.110829
audi              0.098739
opel              0.089460
ford              0.058103
renault           0.037716
peugeot           0.028301
fiat              0.020796
skoda             0.019295
seat              0.017548
smart             0.016866
toyota            0.014792
mazda             0.014464
citroen           0.013891
nissan            0.013837
mini              0.010998
hyundai           0.010916
volvo             0.008897
sonstige_autos    0.007969
kia               0.007805
honda             0.007423
mitsubishi        0.006987
chevrolet         0.006250
alfa_romeo        0.006113
suzuki            0.005813
porsche           0.005486
dacia             0.003330
chrysler          0.003138
jeep              0.002756
land_rover        0.002456
jaguar            0.001829
subaru            0.001747
daihatsu          0.001719
saab              0.001392
daewoo            0.000928
rover             0.000737
l

In [48]:
autos['brand'].value_counts(normalize = True).head(20).sum()

0.9322635227334753

As explored above, the top 20 most common brands represents more than 90% of all listings, so doing aggregation on top20 brands would be a good representation.

In [49]:
top20_brands = autos['brand'].value_counts().head(20).index
top20_brands

Index(['volkswagen', 'bmw', 'mercedes_benz', 'audi', 'opel', 'ford', 'renault',
       'peugeot', 'fiat', 'skoda', 'seat', 'smart', 'toyota', 'mazda',
       'citroen', 'nissan', 'mini', 'hyundai', 'volvo', 'sonstige_autos'],
      dtype='object')

In [50]:
top20_brands_mean_price = {}

for b in top20_brands:
    price_mean = autos.loc[autos['brand'] == b, 'price'].mean()
    top20_brands_mean_price[b] = price_mean 


In [51]:
top20_brands_mean_price

{'audi': 10116.68877833057,
 'bmw': 8932.461820534943,
 'citroen': 4493.8939096267195,
 'fiat': 3923.968503937008,
 'ford': 4908.70032879286,
 'hyundai': 6181.8725,
 'mazda': 5309.526415094339,
 'mercedes_benz': 8890.42748091603,
 'mini': 10724.992555831266,
 'nissan': 6428.428007889546,
 'opel': 4192.410921293472,
 'peugeot': 3948.3760848601737,
 'renault': 3514.8241678726486,
 'seat': 5638.640746500778,
 'skoda': 6841.9066478076375,
 'smart': 3780.4692556634304,
 'sonstige_autos': 11110.243150684932,
 'toyota': 5567.477859778598,
 'volkswagen': 6582.95,
 'volvo': 6075.19018404908}

In [52]:
top20_brand_price = pd.DataFrame(data = [top20_brands_mean_price.values()], columns = top20_brands_mean_price.keys())
top20_brand_price = top20_brand_price.T # Transpose original df
top20_brand_price

0
volvo            6075.190184
hyundai          6181.872500
seat             5638.640747
sonstige_autos  11110.243151
bmw              8932.461821
volkswagen       6582.950000
audi            10116.688778
mazda            5309.526415
skoda            6841.906648
smart            3780.469256
fiat             3923.968504
toyota           5567.477860
nissan           6428.428008
mercedes_benz    8890.427481
opel             4192.410921
mini            10724.992556
citroen          4493.893910
renault          3514.824168
ford             4908.700329
peugeot          3948.376085

In [53]:
top20_brand_price.describe()

0
count     20.000000
mean    6358.172466
std     2386.242367
min     3514.824168
25%     4418.523163
50%     5856.915465
75%     7354.036856
max    11110.243151

In [54]:
top20_brand_price.sort_values(by = 0 , ascending=False) # Sort df by value

0
sonstige_autos  11110.243151
mini            10724.992556
audi            10116.688778
bmw              8932.461821
mercedes_benz    8890.427481
skoda            6841.906648
volkswagen       6582.950000
nissan           6428.428008
hyundai          6181.872500
volvo            6075.190184
seat             5638.640747
toyota           5567.477860
mazda            5309.526415
ford             4908.700329
citroen          4493.893910
opel             4192.410921
peugeot          3948.376085
fiat             3923.968504
smart            3780.469256
renault          3514.824168

As shown above, the average price between the top 20 brands is 6358, the cheapest brand is Renault with an average price 3514. The most expansive brand is Sonsitige_autos with an average price 11110.

### 8/9 Explore top 6 common brands 

In [55]:
top6_brand = autos['brand'].value_counts().head(6).index
top6_brand_avrprice_mileage = {}

for b in top6_brand:
    price_mileage_mean = autos.loc[autos['brand'] == b, ['price', 'odometer_km']].mean()
    top6_brand_avrprice_mileage[b] = price_mileage_mean 

top6_brand_avrprice_mileage
# Odometer_km mean shouldn't be infinite 

{'audi': price          1.011669e+04
 odometer_km             inf
 dtype: float64, 'bmw': price          8932.461821
 odometer_km            inf
 dtype: float64, 'ford': price          4908.700329
 odometer_km            inf
 dtype: float64, 'mercedes_benz': price          8890.427481
 odometer_km            inf
 dtype: float64, 'opel': price          4192.410921
 odometer_km            inf
 dtype: float64, 'volkswagen': price          6582.950000
 odometer_km            inf
 dtype: float64}

In [56]:
autos.loc[autos['brand'] == 'volkswagen', ['odometer_km']].dtypes # Debug: check out odometer_km dtypes

odometer_km    object
dtype: object

In [57]:
autos['odometer_km'] = autos['odometer_km'].astype(int) # Convert odometer_km from str to int

In [58]:
# Redo loop
for b in top6_brand:
    price_mileage_mean = autos.loc[autos['brand'] == b, ['price', 'odometer_km']].mean()
    top6_brand_avrprice_mileage[b] = price_mileage_mean 

top6_brand_avrprice_mileage

{'audi': price           10116.688778
 odometer_km    127787.451631
 dtype: float64, 'bmw': price            8932.461821
 odometer_km    132309.102675
 dtype: float64, 'ford': price            4908.700329
 odometer_km    120807.891029
 dtype: float64, 'mercedes_benz': price            8890.427481
 odometer_km    130667.323319
 dtype: float64, 'opel': price            4192.410921
 odometer_km    124385.295912
 dtype: float64, 'volkswagen': price            6582.95
 odometer_km    126000.00
 dtype: float64}

In [59]:
top6_price_mileage = pd.DataFrame(top6_brand_avrprice_mileage)
top6_price_mileage = top6_price_mileage.T
top6_price_mileage.columns = ['mean_price', 'mean_mileage']

top6_price_mileage.sort_values(by = 'mean_price', ascending=False) # Sort by price high to low

mean_price   mean_mileage
audi           10116.688778  127787.451631
bmw             8932.461821  132309.102675
mercedes_benz   8890.427481  130667.323319
volkswagen      6582.950000  126000.000000
ford            4908.700329  120807.891029
opel            4192.410921  124385.295912

In [60]:
top6_price_mileage.sort_values(by = 'mean_mileage') # Sort by mileage low to high

mean_price   mean_mileage
ford            4908.700329  120807.891029
opel            4192.410921  124385.295912
volkswagen      6582.950000  126000.000000
audi           10116.688778  127787.451631
mercedes_benz   8890.427481  130667.323319
bmw             8932.461821  132309.102675

1. Price of the more expansive cars like Audi, BMW and Mercedes_Benz also don't get effected by mileage much
2. Price of less expensive Ford and Opel are effected by mileage a lot more
3. Volkswagen is in between

### 9/9 Last cleanings

In [61]:
autos.head()

date_crawled                                               name  \
0 2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1 2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2 2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3 2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4 2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type  price   abtest vehicle_type  registration_year  \
0  privat    Angebot   5000  control          bus               2004   
1  privat    Angebot   8500  control    limousine               1997   
2  privat    Angebot   8990     test    limousine               2009   
3  privat    Angebot   4350  control   kleinwagen               2007   
4  privat    Angebot   1350     test        kombi               2003   

     gearbox  power_p_s   model  odometer_km  registration_month fuel_type  \
0    manuell        158  andere       150000                   3       lpg   
1  automatik        286     7er       150000                   6    benzin   
2    manuell        102    golf        70000                   7    benzin   
3  automatik         71  fortwo        70000                   6    benzin   
4    manuell          0   focus       150000                   7    benzin   

        brand unrepaired_damage           ad_created  nr_of_pictures  \
0     peugeot              nein  2016-03-26 00:00:00               0   
1         bmw              nein  2016-04-04 00:00:00               0   
2  volkswagen              nein  2016-03-26 00:00:00               0   
3       smart              nein  2016-03-12 00:00:00               0   
4        ford              nein  2016-04-01 00:00:00               0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

In [62]:
# Drop some of the columns that mostly has only one values: 'seller', 'offer_type', 'nr_of_picture'
autos.drop(['seller','offer_type','nr_of_pictures'], axis = 1, inplace = True)

In [63]:
autos.head()

date_crawled                                               name  \
0 2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1 2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2 2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3 2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4 2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   price   abtest vehicle_type  registration_year    gearbox  power_p_s  \
0   5000  control          bus               2004    manuell        158   
1   8500  control    limousine               1997  automatik        286   
2   8990     test    limousine               2009    manuell        102   
3   4350  control   kleinwagen               2007  automatik         71   
4   1350     test        kombi               2003    manuell          0   

    model  odometer_km  registration_month fuel_type       brand  \
0  andere       150000                   3       lpg     peugeot   
1     7er       150000                   6    benzin         bmw   
2    golf        70000                   7    benzin  volkswagen   
3  fortwo        70000                   6    benzin       smart   
4   focus       150000                   7    benzin        ford   

  unrepaired_damage           ad_created  postal_code            last_seen  
0              nein  2016-03-26 00:00:00        79588  2016-04-06 06:45:54  
1              nein  2016-04-04 00:00:00        71034  2016-04-06 14:45:08  
2              nein  2016-03-26 00:00:00        35394  2016-04-06 20:15:37  
3              nein  2016-03-12 00:00:00        33729  2016-03-15 03:16:28  
4              nein  2016-04-01 00:00:00        39218  2016-04-01 14:38:50

In [64]:
# Columns with german words: gearbox, fuel_type, unrepaired_damage

In [65]:
autos['gearbox'].value_counts()

manuell      26475
automatik     8954
Name: gearbox, dtype: int64

In [66]:
autos['gearbox'] = autos['gearbox'].str.replace('manuell', 'manual')
autos['gearbox'] = autos['gearbox'].str.replace('automatik', 'automatic')
autos['gearbox'].value_counts()

manual       26475
automatic     8954
Name: gearbox, dtype: int64

In [67]:
autos['fuel_type'].value_counts()

benzin     20885
diesel     13314
lpg          556
cng           59
hybrid        37
elektro       14
andere         7
Name: fuel_type, dtype: int64

In [68]:
autos['fuel_type'] = autos['fuel_type'].str.replace('benzin', 'gasoline')
autos['fuel_type'] = autos['fuel_type'].str.replace('elektro', 'electric')
autos['fuel_type'] = autos['fuel_type'].str.replace('andere', 'others')
autos['fuel_type'].value_counts()

gasoline    20885
diesel      13314
lpg           556
cng            59
hybrid         37
electric       14
others          7
Name: fuel_type, dtype: int64

In [69]:
autos['unrepaired_damage'].value_counts()

nein    29290
ja       2287
Name: unrepaired_damage, dtype: int64

In [70]:
autos['unrepaired_damage'] = autos['unrepaired_damage'].str.replace('nein', 'no')
autos['unrepaired_damage'] = autos['unrepaired_damage'].str.replace('ja', 'yes')
autos['unrepaired_damage'].value_counts()

no     29290
yes     2287
Name: unrepaired_damage, dtype: int64

In [71]:
# Convert dates to uniform numeric data. 
#Columns: date_crawled, ad_created, last_seen

In [72]:
autos[['date_crawled', 'ad_created', 'last_seen']].dtypes # Checkout dtypes of the columns 

date_crawled    datetime64[ns]
ad_created              object
last_seen               object
dtype: object

In [73]:
autos['date_crawled'] = autos['date_crawled'].astype(str) # Convert date_crawled back to str

In [74]:
autos['date_crawled'] = autos['date_crawled'].str[0:10]
autos['ad_created'] = autos['ad_created'].str[0:10]
autos['last_seen'] = autos['last_seen'].str[0:10]
autos[['date_crawled', 'ad_created', 'last_seen']].head()

date_crawled  ad_created   last_seen
0   2016-03-26  2016-03-26  2016-04-06
1   2016-04-04  2016-04-04  2016-04-06
2   2016-03-26  2016-03-26  2016-04-06
3   2016-03-12  2016-03-12  2016-03-15
4   2016-04-01  2016-04-01  2016-04-01

In [75]:
autos['date_crawled'] = autos['date_crawled'].str.replace('-','').astype(int)
autos['ad_created'] = autos['ad_created'].str.replace('-','').astype(int)
autos['last_seen'] = autos['last_seen'].str.replace('-','').astype(int)

In [76]:
autos[['date_crawled', 'ad_created', 'last_seen']].head() , autos[['date_crawled', 'ad_created', 'last_seen']].dtypes

(   date_crawled  ad_created  last_seen
 0      20160326    20160326   20160406
 1      20160404    20160404   20160406
 2      20160326    20160326   20160406
 3      20160312    20160312   20160315
 4      20160401    20160401   20160401, date_crawled    int64
 ad_created      int64
 last_seen       int64
 dtype: object)

In [77]:
autos['name'].value_counts()

BMW_318i                                                       67
BMW_320i                                                       67
Volkswagen_Golf_1.4                                            58
BMW_316i                                                       56
BMW_116i                                                       53
Opel_Corsa_1.2_16V                                             53
Volkswagen_Polo                                                49
Volkswagen_Golf                                                48
Volkswagen_Polo_1.2                                            46
BMW_320d_DPF_Touring                                           37
Volkswagen_Golf_1.6                                            37
Volkswagen_Passat_Variant_1.9_TDI                              35
BMW_520i                                                       35
Audi_A4_Avant_2.0_TDI_DPF                                      34
Volkswagen_Golf_1.9_TDI                                        32
Opel_Corsa

In [78]:
brand_model_combo = autos.groupby(['brand','model']).size().reset_index().rename(columns={0:'count'}).sort_values(by = 'count', ascending = False)

In [79]:
# referece: https://stackoverflow.com/questions/35268817/unique-combinations-of-values-in-selected-columns-in-pandas-data-frame-and-countz

In [80]:
# The most common brand/model combinations:
brand_model_combo.head()

brand   model  count
263  volkswagen    golf   2916
23          bmw     3er   2281
267  volkswagen  passat   1152
13         audi      a4   1092
24          bmw     5er   1064

In [81]:
autos['odometer_km'].describe() 

count     36642.000000
mean     123203.018394
std       40325.343994
min        5000.000000
25%      100000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64

In [82]:
autos['odometer_km'].value_counts()

150000    22214
125000     4129
100000     1763
90000      1497
80000      1271
70000      1096
60000      1030
50000       943
40000       750
30000       698
20000       642
5000        414
10000       195
Name: odometer_km, dtype: int64

In [83]:
# Judging from percentile of odometer_km, it's resonable to split it into below groups
odo_150 = autos[autos['odometer_km'] == 150000]
odo_125 = autos[autos['odometer_km'] == 125000]
odo_other = autos[autos['odometer_km'] <= 100000]

In [84]:
odo_150['price'].mean()

4791.420005401998

In [85]:
odo_125['price'].mean()

6958.977476386534

In [86]:
odo_other['price'].mean()

11658.297795902516

How much cheaper are cars with damage than their non-damaged counterparts?

In [87]:
# Find all damaged cars 
damaged_cars = autos.loc[autos['unrepaired_damage'] == 'yes', ['name','price','unrepaired_damage']]
damaged_cars.head()

name  price unrepaired_damage
28                          MINI_Cooper_D   5250               yes
51                Mercedes_Benz_S_320_CDI   6000               yes
81   Nissan_Micra_K12___super_Kleinwagen!   2000               yes
100          Mazda_5_2.0_CD_DPF_Exclusive   4500               yes
110                           Ford_Galaxy   1300               yes

In [88]:
# Find all undamaged cars 
undamaged_cars = autos.loc[autos['unrepaired_damage'] == 'no', ['name','price','unrepaired_damage']]
undamaged_cars.head()

name  price unrepaired_damage
0                   Peugeot_807_160_NAVTECH_ON_BOARD   5000                no
1         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   8500                no
2                         Volkswagen_Golf_1.6_United   8990                no
3  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   4350                no
4  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   1350                no

In [89]:
# Concatenate damaged cars and undamaged cars
price_damage_correlation = pd.concat([damaged_cars, undamaged_cars], axis = 0)
price_damage_correlation

name  price  \
28                                         MINI_Cooper_D   5250   
51                               Mercedes_Benz_S_320_CDI   6000   
81                  Nissan_Micra_K12___super_Kleinwagen!   2000   
100                         Mazda_5_2.0_CD_DPF_Exclusive   4500   
110                                          Ford_Galaxy   1300   
115                           Audi_A3_1.8T_S_LINE_179_PS   2300   
136                      2_mal_Subaru_Libero_super_delux   1499   
161                    Volkswagen_Touareg_3.0_V6_TDI_DPF  11000   
182    Renault_Megane_2.0_Coupe_Cabriolet_Aut._Privilege   1790   
198     Fiat_Ducato_Maxi_Lang_L5.___3_Liter_Motor_158_ps   4200   
216                    Golf_4___1.6sr_Motor_zu_verkaufen   1600   
279                                   Opel_Corsa_1.0_12V   1500   
352       Ford_Focus_LPG_Turnier_1.6_16V_Titanium_DEFEKT   1500   
377                               VW_Golf_5_Goal_Edition   4700   
396                             2_5_Liter_Caravelle_lang   5500   
397                     Mercedes_Benz_A_170_CDI_Elegance   1150   
422                               Mercedes_Benz_Sprinter   4900   
428                                           Audi_A3_98   1000   
431                                       Seat_Ibiza_Rot   1300   
432                                     BMW_X6_xDrive35d  23900   
440                VW_Golf_4_1.6_4_Tuerer_HU_AU_sind_neu   1000   
463                                    Audi_A4_Avant_1.6   2000   
494                                   Mazda_6__2.0Diesel   1200   
506                                         Kia_Carneval   1800   
508                          Renault_Twingo_1.2_16V_Aut.   1400   
572                  Volkswagen_Golf_Plus_1.9_TDI_United   4000   
580                                             BMW_520i   1400   
695                               Volkswagen_Kastenwagen  18000   
708                                   Honda_Jazz_1_4__ES   2700   
725                                           VW_Polo_6N   1111   
...                                                  ...    ...   
49947  Audi_A5_2.7_TDI_+S_Line_Vollausstattung/ABT/B&...  23600   
49948                Hyundai_ix35_2.0_CRDi_4WD_Automatik  19900   
49950         Volvo_V70_2_II__2002__AHK__Klima__Volleder   3099   
49951  Opel_Corsa_D1.0_weiss_/Scheckheftgepflegt/_Unf...   3550   
49952  VW_POLO__9N3_Silber_1_2_Ltr._5_tuerer__96_tsd....   4300   
49955                     Golf_1.9_TDI_DPF_4mot_GT_Sport   3400   
49956        Toyota_Yaris_1.3_VVT_i_Executive_MODEL_2007   4450   
49959           Mercedes_Benz_C_180_T_Kompressor_Classic   4200   
49963            Mercedes_Benz_B_200_CDI_Special_Edition   7600   
49965          Opel_Astra_1.6_Lenkradheizung~Sitzheizung   6700   
49968                  Mercedes_Benz_190_D_2.5_Automatik   2100   
49969             Nissan_X_Trail_2.2_dCi_4x4_Sport_m.AHZ   4500   
49970  c4_Grand_Picasso_mit_Automatik_Leder_Navi_Temp...  15800   
49973                   Mercedes_Benz_SLK_200_Kompressor   6000   
49975          Honda_Jazz_1.3_DSi_i_VTEC_IMA_CVT_Comfort   9700   
49976  Audi_80_Avant_2.6_E__Vollausstattung!!_Einziga...   5900   
49977                        Mercedes_Benz_C200_Cdi_W203   5500   
49979                      Volkswagen_Polo_1.6_TDI_Style  11000   
49981             Opel_Astra_Kombi_mit_Anhaengerkupplung   2000   
49986  Chrysler_300C_3.0_CRD_DPF_Automatik_Voll_Ausst...  15900   
49987  Audi_A3_Limousine_2.0_TDI_DPF_Ambition__NAVI__...  21990   
49988                                         BMW_330_Ci   9550   
49990         Mercedes_Benz_A_200__BlueEFFICIENCY__Urban  17500   
49992              Fiat_Grande_Punto_1.4_T_Jet_16V_Sport   4800   
49994  Audi_A6__S6__Avant_4.2_quattro_eventuell_Tausc...   5000   
49995   Audi_Q5_3.0_TDI_qu._S_tr.__Navi__Panorama__Xenon  24900   
49996  Opel_Astra_F_Cabrio_Bertone_Edition___TÜV_neu+...   1980   
49997                     Fiat_500_C_1.2_Dualogic_Lounge  13200   
49998                 Audi_A3_2.

In [90]:
# Use duplicated() with subset(), se keep=False to drop non-duplicated('name') rows(drop unique cars)
price_damage_correlation = price_damage_correlation[price_damage_correlation.duplicated(subset='name', keep=False)]
price_damage_correlation

name  price  \
28                                         MINI_Cooper_D   5250   
51                               Mercedes_Benz_S_320_CDI   6000   
100                         Mazda_5_2.0_CD_DPF_Exclusive   4500   
110                                          Ford_Galaxy   1300   
161                    Volkswagen_Touareg_3.0_V6_TDI_DPF  11000   
279                                   Opel_Corsa_1.0_12V   1500   
397                     Mercedes_Benz_A_170_CDI_Elegance   1150   
422                               Mercedes_Benz_Sprinter   4900   
432                                     BMW_X6_xDrive35d  23900   
463                                    Audi_A4_Avant_1.6   2000   
580                                             BMW_520i   1400   
725                                           VW_Polo_6N   1111   
777                                   Seat_toledo_Unfall   1300   
782                Volkswagen_Golf_1.6_Automatik_Edition   3350   
798                                 BMW_325d_DPF_Touring   6100   
883                         Renault_Megane_1.6_Grandtour   1800   
1150                               Chevrolet_Aveo_1.2_LS   1580   
1188                                   Mercedes_A_Klasse   1200   
1434                        Peugeot_807_HDi_130_Tendance   1000   
1441                                     Audi_A3_2.0_FSI   6200   
1513                  Skoda_Fabia_1.2_HTP_Combi_Ambiente   4400   
1621                 Audi_A3_1.4_TFSI_Sportback_Ambition   8500   
1668                                    Toyota_Aygo_Cool   1850   
1681                        Mazda_5_2.0_CD_DPF_Exclusive   2800   
1828                    Mercedes_Benz_V_220_CDI_Ambiente   2800   
1862                                         MG_MGF_1.8i   1300   
1892       Mercedes_Benz_ML_350_CDI_4Matic_7G_TRONIC_DPF  15990   
1952   Mercedes_Benz_E_350_CDI_DPF_BlueEFFICIENCY_7G_...  15600   
1983             Volkswagen_Golf_Variant_1.4_Comfortline   3700   
1986                                     Audi_A4_2.5_TDI   3620   
...                                                  ...    ...   
49877                   Mercedes_Benz_C_200_T_Kompressor   4550   
49883                                            BMW_316   1500   
49893       Mitsubishi_Pajero_3.2_DI_D_Automatik_Instyle  19500   
49894                      Mercedes_Benz_E_200_T_Classic   1500   
49897                                           BMW_523i   1890   
49901                                      MINI_Mini_One   8300   
49903  Opel_Astra_1.7_CDTI_Caravan_DPF__119g__Edition...   8250   
49907                     Opel_Astra_1.4_ecoFLEX_Edition   8250   
49908                              Fiat_Panda_1.1_Active   2900   
49913          Mercedes_Benz_C_200_Kompressor_Sportcoupe   2500   
49915                                 Renault_Twingo_1.2   1700   
49922                        Ford_Focus_2.0_16V_Titanium   2888   
49923                             Toyota_Yaris_1.0_VVT_i   4500   
49929                                   Opel_Corsa_Swing   1390   
49930  Opel_Insignia_2.0_CDTI_Sports_Tourer_Automatik...   8900   
49933                      Volkswagen_New_Beetle_1.9_TDI   5500   
49938                     Mercedes_Benz_A_160_Avantgarde   2300   
49939                          Audi_TT_Roadster_2.0_TFSI  16995   
49942             Audi_A3_2.0_TDI_Sportback_DPF_Ambition  10000   
49945                  omega_2_2_facelift_mit_neuem_tuev   1750   
49948                Hyundai_ix35_2.0_CRDi_4WD_Automatik  19900   
49955                     Golf_1.9_TDI_DPF_4mot_GT_Sport   3400   
49959           Mercedes_Benz_C_180_T_Kompressor_Classic   4200   
49973                   Mercedes_Benz_SLK_200_Kompressor   6000   
49979                      Volkswagen_Polo_1.6_TDI_Style  11000   
49987  Audi_A3_Limousine_2.0_TDI_DPF_Ambition__NAVI__...  21990   
49988                                         BMW_330_Ci   9550   
49990         Mercedes_Benz_A_200__BlueEFFICIENCY__Urban  17500   
49992              Fiat_Grande_P

In [91]:
# Find indexis of cars with only one damage condition(only damaged or undamaged) and drop them 
names = price_damage_correlation['name'].unique()
to_drop = []
target_ix = []
len(names)

2753

In [92]:
for n in names: 
    if price_damage_correlation.loc[price_damage_correlation['name'] == n, 'unrepaired_damage'].unique().size<2:
        target_ix = price_damage_correlation[price_damage_correlation['name'] == n].index
        to_drop.extend(target_ix)

In [93]:
to_drop

[777,
 32299,
 1150,
 17480,
 2962,
 40745,
 3603,
 9062,
 6197,
 7091,
 6880,
 22871,
 8972,
 12253,
 12265,
 38927,
 12282,
 21384,
 13343,
 26378,
 13411,
 42664,
 17613,
 26221,
 23743,
 39647,
 23987,
 44379,
 24979,
 30033,
 30618,
 37723,
 35590,
 42217,
 2,
 12072,
 14309,
 30536,
 22,
 3966,
 10533,
 15847,
 17183,
 17191,
 18708,
 27906,
 30679,
 34553,
 41232,
 43743,
 45468,
 38,
 9780,
 24319,
 29890,
 37894,
 47,
 1820,
 17686,
 17898,
 19325,
 23874,
 24099,
 34101,
 48047,
 49,
 48152,
 54,
 12699,
 57,
 357,
 9027,
 15616,
 22138,
 30955,
 31596,
 31705,
 32611,
 35052,
 72,
 20079,
 30420,
 48124,
 91,
 5628,
 13833,
 19566,
 23724,
 28006,
 28375,
 36198,
 41210,
 43708,
 93,
 29186,
 104,
 14522,
 108,
 4479,
 39014,
 43451,
 49842,
 109,
 4370,
 15261,
 15757,
 114,
 11412,
 14129,
 23371,
 28048,
 44784,
 116,
 1364,
 3820,
 12913,
 13791,
 16192,
 16206,
 17884,
 18215,
 19064,
 25885,
 26935,
 28414,
 32633,
 33158,
 33915,
 39259,
 41988,
 119,
 26758,
 122,
 2

In [100]:
price_damage_correlation.drop(index = to_drop, axis = 0, inplace = True)

ValueError: labels [  777 32299  1150 ... 48703 49147 49243] not contained in axis

In [121]:
# Compare mean of damaged and undamaged counterpart for each car
price_comparison = {}
compared_cars = price_damage_correlation['name'].unique()
target = None
damaged_mean = None
undamaged_mean = None
len(compared_cars)

531

In [123]:
for c in compared_cars:
    target = price_damage_correlation[price_damage_correlation['name'] == c]
    damaged_mean = target.loc[target['unrepaired_damage']=='yes', 'price'].mean()
    undamaged_mean = target.loc[target['unrepaired_damage']=='no', 'price'].mean()
    price_comparison[c] = undamaged_mean-damaged_mean

In [124]:
price_comparison

{'Alfa_Romeo_147_1.6_Twin_Spark_Distinctive': 383.79999999999995,
 'Alfa_Romeo_156_2.0_JTS': -1700.0,
 'Audi_80_Avant_2.6_E': 3900.0,
 'Audi_A3': 2259.9285714285716,
 'Audi_A3_1.4_TFSI_Sportback_Ambition': 5649.799999999999,
 'Audi_A3_1.6_Ambiente': 752.5714285714284,
 'Audi_A3_1.6_Sportback_Ambiente': 2766.333333333334,
 'Audi_A3_1.8': 27.77777777777783,
 'Audi_A3_1.8_Ambiente': 392.1666666666665,
 'Audi_A3_1.8_Attraction': -650.0,
 'Audi_A3_1.9_TDI': 917.375,
 'Audi_A3_1.9_TDI_Ambiente': 2830.0,
 'Audi_A3_1.9_TDI_DPF_Ambition': 1300.0,
 'Audi_A3_1.9_TDI_DPF_Attraction': 1900.0,
 'Audi_A3_2.0_FSI': -450.5,
 'Audi_A3_2.0_FSI_Ambiente': 570.0,
 'Audi_A3_2.0_FSI_Ambition': 1490.0,
 'Audi_A3_2.0_TDI_DPF': 400.0,
 'Audi_A3_2.0_TDI_Sportback_Ambition': 9116.666666666668,
 'Audi_A4': 2097.6000000000004,
 'Audi_A4_1.6': 1424.4545454545455,
 'Audi_A4_1.8_T': 2762.7142857142853,
 'Audi_A4_1.9_TDI': 1531.7272727272725,
 'Audi_A4_2.0_TDI_DPF': 5097.5,
 'Audi_A4_2.5_TDI': 1487.25,
 'Audi_A4_Avant'

In [126]:
price_comparison = pd.Series(price_comparison)
price_comparison

Alfa_Romeo_147_1.6_Twin_Spark_Distinctive                 383.800000
Alfa_Romeo_156_2.0_JTS                                  -1700.000000
Audi_80_Avant_2.6_E                                      3900.000000
Audi_A3                                                  2259.928571
Audi_A3_1.4_TFSI_Sportback_Ambition                      5649.800000
Audi_A3_1.6_Ambiente                                      752.571429
Audi_A3_1.6_Sportback_Ambiente                           2766.333333
Audi_A3_1.8                                                27.777778
Audi_A3_1.8_Ambiente                                      392.166667
Audi_A3_1.8_Attraction                                   -650.000000
Audi_A3_1.9_TDI                                           917.375000
Audi_A3_1.9_TDI_Ambiente                                 2830.000000
Audi_A3_1.9_TDI_DPF_Ambition                             1300.000000
Audi_A3_1.9_TDI_DPF_Attraction                           1900.000000
Audi_A3_2.0_FSI                   